### PR TITLE
QUIC for 3.3.0

### PR DESCRIPTION
**Pull request by wbl**
Review carefully: we've tested the code decently well (thanks @nibanks). The branch wladd/quic-comparison-branch has everything squashed down to ease review. I renamed our feature to boring-quic-api in this branch as the first big change, hopefully everywhere it needed to happen in the build system.

### CODE REVIEW

**Comment by wbl** [comment]
Looks like some more work needed for CI to work, will update in a bit.

**Comment by wbl** [comment]
Looks like all checks pass modulo the fuzzer, but that seems to have run fine other times?

**Comment by wbl** [comment]
I do not understand what the difference is between the two fuzzer builds. @tmshort any ideas?

**Comment by tmshort** [comment]
i wouldn't worry about it. The one tagged `(pull_request)` is likely only run on an actual pull request, and is more comprehensive than the one tagged (`push`) which would happen on all pushes to the repo (even if they don't become a PR).

**Comment by tmshort** [comment]
The `CIFuzz (pull_request)` has almost always failed.

**Comment by tmshort** [comment]
Any updates here @wbl ?

**Comment by wbl** [comment]
Sorry I think balls ended up on floors as Ive forgotten to poke. I've made the comparison branch and gotten this one to what I think is reviewable state but another set of eyes would be helpful. So take a look or reremind me what I need to do first.

File: apps/info.c
Comment (richsalz) [inline apps/info.c:13]: does this need an ifdef?

File: ssl/ssl_lib.c
Comment (richsalz) [inline ssl/ssl_lib.c:571]: blank line after decl?

File: ssl/ssl_lib.c
Comment (richsalz) [inline ssl/ssl_lib.c:582]: could merge with the assignment on 584

File: ssl/ssl_lib.c
Comment (richsalz) [inline ssl/ssl_lib.c:1518]: could also merge here.

File: ssl/statem/statem_lib.c
Comment (richsalz) [inline ssl/statem/statem_lib.c:101]: s/sent/send/ s/sice/since/

**Review by richsalz — APPROVED** [review]
some nits, some politics.  didn't look closely at the bulk of the QIUC changes since they look like they always do.

**Comment by wbl** [comment]
Thanks @richsalz for looking it over. I'll fix up the comments if no one else chimes in with more soon in a separate commit.

File: apps/info.c
Comment (wbl) [reply apps/info.c:13]: No, compiles fine in all combinations supported

**Comment by nibanks** [comment]
Sorry, I haven't followed up here in a while. I just updated https://github.com/microsoft/msquic/pull/4274 to use the latest commit here with the latest version of MsQuic and I am seeing build failures on Ubuntu 24.04. Note, the last time I tried this, we didn't build for 24.04 yet, so this isn't necessarily a recent break, but it is a regression from previous v3 openssl.

https://github.com/microsoft/msquic/actions/runs/9974479617/job/27562154754#step:8:2184
```
/usr/bin/ld: ../../../../_deps/opensslquic-build/openssl3/lib/libssl.a(libssl-lib-ssl_lib.o): in function `SSL_CTX_free':
ssl_lib.c:(.text+0x93f1): undefined reference to `OSSL_STACK_OF_X509_free'
/usr/bin/ld: ../../../../_deps/opensslquic-build/openssl3/lib/libssl.a(libssl-lib-ssl_lib.o): in function `SSL_dup':
ssl_lib.c:(.text+0x9e44): undefined reference to `OSSL_STACK_OF_X509_free'
/usr/bin/ld: ../../../../_deps/opensslquic-build/openssl3/lib/libssl.a(libssl-lib-ssl_lib.o): in function `ossl_ssl_connection_free':
ssl_lib.c:(.text+0xa262): undefined reference to `OSSL_STACK_OF_X509_free'
/usr/bin/ld: ssl_lib.c:(.text+0xa540): undefined reference to `OSSL_STACK_OF_X509_free'
/usr/bin/ld: ../../../../_deps/opensslquic-build/openssl3/lib/libssl.a(libssl-lib-ssl_lib.o): in function `SSL_CTX_new_ex':
ssl_lib.c:(.text+0xa6fc): undefined reference to `OPENSSL_LH_set_thunks'
/u
...
```

**Comment by wbl** [comment]
@nibanks I'll look into that failure: if those things got redefined to be macros or functions across versions, maybe its a build picking up wrong things sort of issue, but we should know that for sure and I'll go attack it.  What's the easiest way for me to locally reproduce where I can poke around the image/reproduce in CI: open a junk pr to MSQuic?

**Comment by nibanks** [comment]
> @nibanks I'll look into that failure: if those things got redefined to be macros or functions across versions, maybe its a build picking up wrong things sort of issue, but we should know that for sure and I'll go attack it. What's the easiest way for me to locally reproduce where I can poke around the image/reproduce in CI: open a junk pr to MSQuic?

Do you want to try to repro locally or just let MsQuic CI/CD do the heavy lifting? If you want to use MsQuic, yeah, just make a draft PR and I will approve the automation to run.

**Comment by nibanks** [comment]
> I think I've got to play with it locally just to understand. But wait,
> 
> > cd /__w/msquic/msquic/build/linux/x86_openssl3/src/tools/interopserver && /usr/bin/cmake -E cmake_link_script CMakeFiles/quicinteropserver.dir/link.txt --verbose=1
> > /usr/bin/c++ --sysroot=/ -Og -fno-omit-frame-pointer -ggdb3 CMakeFiles/quicinteropserver.dir/InteropServer.cpp.o -o /__w/msquic/msquic/artifacts/bin/linux/x86_Debug_openssl3/quicinteropserver  -Wl,-rpath,"$ORIGIN:__w/msquic/msquic/artifacts/bin/linux/x86_Debug_openssl3" /__w/msquic/msquic/artifacts/bin/linux/x86_Debug_openssl3/libmsquic.so.2.4.0 ../../../obj/Debug/libplatform.a ../../../obj/Debug/liblogging.a ../../../_deps/opensslquic-build/openssl3/lib/libssl.a /usr/lib/x86_64-linux-gnu/libcrypto.so -ldl /usr/lib/x86_64-linux-gnu/libatomic.so.1 /usr/lib/x86_64-linux-gnu/libnuma.so.1
> 
> Is it just me or is it linking the new libssl.a to the system libcrypto.so? Then I'm not surprised it doesn't work, as those got introduced in patch [openssl@5c42ced](https://github.com/openssl/openssl/commit/5c42ced0ff974a59af98b75e54136f4282718266)

Ah, you're right! It's all our builds with the `-UseSystemOpenSSLCrypto` flag, which statically links libssl, but dynamically links libcrytpo (for FIPS). This passed on previous versions/commits, but is failing now. How are we supposed to handle this?

**Comment by wbl** [comment]
My first reaction is why on earth would you expect this to work out nicely, and why do you need it: is there some other way to get what you want here? Note that the fips module is entirely unaffected by this issue as that dynamically loads with a much more restricted interface.

**Comment by wbl** [comment]
Alternatively try 3.3 as the system version, but that might have other issues

**Comment by nibanks** [comment]
> My first reaction is why on earth would you expect this to work out nicely, and why do you need it: is there some other way to get what you want here? Note that the fips module is entirely unaffected by this issue as that dynamically loads with a much more restricted interface.

We need to dynamically link to the system libcrypto for FIPS compliance (because we can't bring our own binary). But we cannot link to the system libssl because it's not the quictls fork. This has worked out so far for previous v1 and v3 versions of openssl.

> Alternatively try 3.3 as the system version, but that might have other issues

I don't think we had to specify anything version specific in the past. Any suggestions on how we'd test this out?

**Comment by wbl** [comment]
> > My first reaction is why on earth would you expect this to work out nicely, and why do you need it: is there some other way to get what you want here? Note that the fips module is entirely unaffected by this issue as that dynamically loads with a much more restricted interface.
> 
> We need to dynamically link to the system libcrypto for FIPS compliance (because we can't bring our own binary). But we cannot link to the system libssl because it's not the quictls fork. This has worked out so far for previous v1 and v3 versions of openssl.

That's odd: they were supposed to have limited fips issues to fips.so, but maybe you're doing something different.
> 
> > Alternatively try 3.3 as the system version, but that might have other issues
> 
> I don't think we had to specify anything version specific in the past. Any suggestions on how we'd test this out?

Ideally just an apt-get update and install the new version ahead of running the existing script. Not really sure about where it is in ubuntu.

**Comment by tmshort** [comment]
What's the status of this PR?

**Comment by wbl** [comment]
I'd like more than just Rich's approval on it, given the size and scope and the number of decisions I had to make that people could possibly have issues with. Please take a look and either approve or say what you want to change.

**Comment by tatsuhiro-t** [comment]
Built fine with the latest ngtcp2 and all tests passed.  Thank you for great work.
Once thing, that is I do not know whether it is the issue of original openssl, but the generated pkgconfig file is broken.  libdir lacks lib directory.

**Comment by nibanks** [comment]
I just pushed the latest to https://github.com/microsoft/msquic/pull/4274 to run it all again. If only our 'use system libcrypto' tests fail, I'm fine for merging this. I still need to figure out a solution for that though...

**Comment by richsalz** [comment]
@nibanks I agree, let's merge this once your results are known.

**Comment by wbl** [comment]
@nibanks it looks like it's just the usesystem openssl tests failing, but I'm not that familiar with your tests.

@tatsuhiro-t pkgconfig is a bit of a black box to me I'm afraid. If you would be so kind, open another issue after we merge this where we can solve that problem.

**Comment by nibanks** [comment]
I'm sorry. I just haven't had the time to follow up on this. I will see if I can make progress tomorrow morning, otherwise, don't wait on me.

**Comment by tatsuhiro-t** [comment]
@wbl All right. Lets move on.  Will check after the merge to check that it is still an issue.

**Comment by rapsys** [comment]
I have two questions related to this merge, not sure it's the best place to ask, sorry in advance.

Would it be possible to get a release with quic patchset on top of openssl-3.3.2, it has two CVE fixed in it ?

Do we need to build this quictls with enable-quic & no-boring-ssl-api ?

I tried applying openssl 3.3.0->3.3.2 patchset on top of quictls-3.3.0 but when enabling quic with boring ssl api disable, I get these errors:
```
Test Summary Report
-------------------
25-test_verify.t                      (Wstat: 512 (exited 2) Tests: 193 Failed: 2)
  Failed tests:  127, 129
  Non-zero exit status: 2
70-test_quic_multistream.t            (Wstat: 256 (exited 1) Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
70-test_quic_tserver.t                (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
75-test_quicapi.t                     (Wstat: 512 (exited 2) Tests: 2 Failed: 2)
  Failed tests:  1-2
  Non-zero exit status: 2
90-test_quicfaults.t                  (Wstat: 512 (exited 2) Tests: 2 Failed: 2)
  Failed tests:  1-2
  Non-zero exit status: 2
90-test_sslapi.t                      (Wstat: 512 (exited 2) Tests: 4 Failed: 2)
  Failed tests:  1, 3
  Non-zero exit status: 2
Files=314, Tests=2882, 436 wallclock secs ( 7.35 usr  0.40 sys + 358.88 cusr 35.17 csys = 401.80 CPU)
Result: FAIL
```

**Comment by richsalz** [comment]
It has been merged, but we're no longer actively working on 3.3+quic.  We are instead focused on the quictls/quictls fork.

If the community wants to take over maintenance of this branch, feel free.  We can add committers as needed.

**Pull request merged**
